### ISSUE TITLE
Layer object

### ISSUE THREAD

**Issue by MarcCote** [SMART-Lab/smartlearner#20]
I still think we should have `Layer` objects that we could manipulate to form the model network.

This is what I have in mind for a simple feed forward neural network with two hidden layers.

``` python
input_layer = Layer(size=trainset.input_size)
first_hidden_layer = FullyConnectedLayer(size=100, activation_fct=SigmoidActivation)
second_hidden_layer = FullyConnectedLayer(size=50, activation_fct=TanhActivation)
output_layer = FullyConnectedLayer(size=trainset.target_size, activation_fct=SigmoidActivation)

network = input_layer + first_hidden_layer + second_hidden_layer + output_layer
model = NN(network, ...)
```

This is an example for an RNN with two inputs.

``` python
input_layer = Layer(size=trainset.input_size)
first_hidden_layer = RecurrentLayer(size=100, activation_fct=SigmoidActivation)
second_hidden_layer = RecurrentLayer(size=50, activation_fct=TanhActivation)
another_hidden_layer = RecurrentLayer(size=50, activation_fct=ReLuActivation)
output_layer = FullyConnectedLayer(size=trainset.target_size, activation_fct=SigmoidActivation)

network1 = input_layer + first_hidden_layer + second_hidden_layer
network2 = input_layer + another_hidden_layer
merged_networks = AggregateNetworks(network1, network2, method="sum")
network = merged_networks + output_layer
model = RNN(network, ...)
```

`network.parameters` should provide a list of the parameters of every layers.

More thinking is needed to support dropout and batch normalization, maybe some sort of decorator design pattern would be useful for that.

More thinking is needed for models that have multiple outputs. @havaeimo I would love to have your input for this.

**Comment by ASalvail** [comment]
Who manages the weights?
If it's the network, alright, though I don't really see what is different from what we've currently have in models.

If it's the layers, you'd have to make a graph out of them. Not impossible, but really messy when adding recurrence.

**Comment by MarcCote** [comment]
The layer receiving the connections and have an activation funciton. Basically, only `input_layer` won't have weights to manage.

**Comment by MarcCote** [comment]
This has been moved over here https://github.com/SMART-Lab/models/issues/10
